### PR TITLE
feat(*): add functions for merging traffic policies

### PR DESCRIPTION
**Pull request by michelleN**
part of #2034

Signed-off-by: Michelle Noorali <michellemolu@gmail.com>

**Description**:

**Affected area**:

- New Functionality      [ ]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [ ]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

### CODE REVIEW

File: pkg/trafficpolicy/trafficpolicy.go [hunk 1]
```diff
@@ -56,3 +56,81 @@ func (out *OutboundTrafficPolicy) AddRoute(httpRouteMatch HTTPRouteMatch, weight
 	}
 	return nil
 }
+
+// MergeInboundPolicies merges latest InboundTrafficPolicies into original InboundTrafficPolicies
```
Comment (shashankram) [inline pkg/trafficpolicy/trafficpolicy.go:60]: The term `original` is a bit unclear to me. Maybe another line on what `original` refers to here would be helpful.

**Review by shashankram — APPROVED** [review]

File: pkg/trafficpolicy/trafficpolicy.go [hunk 2]
```diff
@@ -56,3 +56,81 @@ func (out *OutboundTrafficPolicy) AddRoute(httpRouteMatch HTTPRouteMatch, weight
...
+	for _, l := range latest {
+		foundHostnames := false
+		for _, or := range original {
+			if reflect.DeepEqual(or.Hostnames, l.Hostnames) {
+				foundHostnames = true
+				or.Rules = mergeRules(or.Rules, l.Rules)
+			}
+		}
+		if !foundHostnames {
+			original = append(original, l)
```
Comment (draychev) [inline pkg/trafficpolicy/trafficpolicy.go:71]: I suspect we may end up with duplicate `InboundTrafficPolicy`? Would a map/set make sense?

**Review by draychev — APPROVED** [review]

File: pkg/trafficpolicy/trafficpolicy.go (hunk 2 above)
Comment (michelleN) [reply pkg/trafficpolicy/trafficpolicy.go:71]: Yeah let me look into that. I've been a bit worries about passing sets around bc you can't specify the type of objects in the set when you're passing if to and from methods.

File: pkg/trafficpolicy/trafficpolicy.go (hunk 1 above)
Comment (ksubrmnn) [reply pkg/trafficpolicy/trafficpolicy.go:60]: Agreed

File: pkg/trafficpolicy/trafficpolicy.go [hunk 3]
```diff
@@ -56,3 +56,81 @@ func (out *OutboundTrafficPolicy) AddRoute(httpRouteMatch HTTPRouteMatch, weight
 	}
 	return nil
 }
+
+// MergeInboundPolicies merges latest InboundTrafficPolicies into original InboundTrafficPolicies
+func MergeInboundPolicies(original []*InboundTrafficPolicy, latest ...*InboundTrafficPolicy) []*InboundTrafficPolicy {
+	for _, l := range latest {
+		foundHostnames := false
+		for _, or := range original {
+			if reflect.DeepEqual(or.Hostnames, l.Hostnames) {
```
Comment (ksubrmnn) [inline pkg/trafficpolicy/trafficpolicy.go:65]: Is there a scenario where the hostnames would overlap but not be equal? Is the list of hostnames guaranteed to always be a complete set? If so, how?

File: pkg/trafficpolicy/trafficpolicy.go [hunk 4]
```diff
@@ -56,3 +56,81 @@ func (out *OutboundTrafficPolicy) AddRoute(httpRouteMatch HTTPRouteMatch, weight
...
+				mergedRoutes, errs := mergeRoutesWeightedClusters(or.Routes, l.Routes)
+				or.Routes = mergedRoutes
+				routesErrors = append(routesErrors, errs...)
+			}
+		}
+		if !foundHostnames {
+			original = append(original, l)
+		}
+	}
+	return original, routesErrors
```
Comment (ksubrmnn) [inline pkg/trafficpolicy/trafficpolicy.go:94]: should original be nil if routesErrors is not nil?

File: pkg/trafficpolicy/trafficpolicy.go [hunk 5]
```diff
@@ -56,3 +56,81 @@ func (out *OutboundTrafficPolicy) AddRoute(httpRouteMatch HTTPRouteMatch, weight
...
+//	for any HTTPRouteMatch
+func mergeRoutesWeightedClusters(originalRoutes, latestRoutes []*RouteWeightedClusters) ([]*RouteWeightedClusters, []error) {
+	mergeErrors := []error{}
+	for _, latest := range latestRoutes {
+		foundRoute := false
+		for _, original := range originalRoutes {
+			if reflect.DeepEqual(original.HTTPRouteMatch, latest.HTTPRouteMatch) {
+				foundRoute = true
+				if !reflect.DeepEqual(original.WeightedClusters, latest.WeightedClusters) {
+					mergeErrors = append(mergeErrors, errors.Errorf("Error merging RoutesWeightedClusters %v with %v", original, latest))
```
Comment (ksubrmnn) [inline pkg/trafficpolicy/trafficpolicy.go:126]: I'd change this error msg to be more descriptive

Comment (ksubrmnn) [reply pkg/trafficpolicy/trafficpolicy.go:126]: Why is this an error? What guarantees that the weighted clusters are a complete set?

**Review by ksubrmnn — APPROVED** [review]
Had a few questions. Maybe some more comments can be added in future PRs but approving this now to unblock

**Pull request merged**